### PR TITLE
Provide a way to allow bad HTTP2 cipher suites for client

### LINKED ISSUE #3157: Armeria client can't connect to the server which doesn't support modern cipher suites
It seems Armeria client doesn’t have the way to connect servers which only supports ciphers in `HTTP2_BLACKLISTED_CIPHERS` even if on HTTP/1.1 connection.
https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/internal/common/util/SslContextUtil.java#L187

I want to add acceptable cipher suites on client using `ClientFactory.builder().tlsCustomizer()`, because some ASP.NET server doesn’t support modern cipher suites (and HTTP/2). But Armeria rejected HTTP/2 blacklisted cipher suites and throw exception on `SslContextUtil.createSslContext()`. I think this restriction is too strict for clients.
https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/internal/common/util/SslContextUtil.java#L178

About the HTTP/2 cipher suite blacklist, RFC7540 says

> An HTTP/2 implementation MAY treat the negotiation of any of the following cipher suites with TLS 1.2 as a connection error (Section 5.4.1) of type INADEQUATE_SECURITY:

but it’s not MUST or REQUIRED.
Probably the target of this blacklist is for server. I don’t think we need too strict cipher suite filtering on client side.
It's OK with filtering them on default value, but I think no need to throw exception for customized value, isn’t it?

Armeria should have a option to add blacklisted cipher suite for compatibility.

Additionally, `tlsNoVerify()` is too insecure for our application.

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

A client fails to connect to a server using HTTPS if the server does not support modern cipher suites.
Because Armeria does not allow to use HTTP/2 cipher suite blacklist if a protocol is not `H1`.
https://github.com/line/armeria/blob/9e0a4c02fdeeb276602c05992396605848d8f843/core/src/main/java/com/linecorp/armeria/internal/common/util/SslContextUtil.java#L133-L135

Modifications:

- Add `tlsAllowUnsafeCiphers` options to ClientFactoryBuilder and Flags
  for disabling the cipher suites validation.

Result:

- You can now use [https://tools.ietf.org/html/rfc7540#appendix-A](TLS 1.2 Cipher Suite Block List) for TLS Handshake.
- Fixes #3157

### CODE REVIEW

**Review by minwoox — APPROVED** [review]
Thanks! 👍

File: core/src/main/java/com/linecorp/armeria/internal/common/util/SslContextUtil.java [hunk 1]
```diff
@@ -184,7 +187,8 @@ private static void validateHttp2Ciphers(Set<String> ciphers) {
     }
 
     // https://httpwg.org/specs/rfc7540.html#BadCipherSuites
-    private static final Set<String> HTTP2_BLACKLISTED_CIPHERS =
+    @VisibleForTesting
+    static final Set<String> HTTP2_BLACKLISTED_CIPHERS =
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/internal/common/util/SslContextUtil.java:191]: `BAD_HTTP2_CIPHERS`

File: core/src/test/java/com/linecorp/armeria/internal/common/util/SslContextUtilTest.java [hunk 2]
```diff
@@ -48,4 +59,51 @@ void jdkSsl() {
...
+        final ClientFactory factory =
+                ClientFactory.builder()
+                             .tlsAllowUnsafeCiphers(true)
+                             .tlsCustomizer(builder -> builder.ciphers(ImmutableList.of(cipher)))
+                             .build();
+        factory.closeAsync();
+    }
+
+    @Nullable
+    private static String getBlackListedCipher() {
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/internal/common/util/SslContextUtilTest.java:86]: `getBadCipher()`?

File: core/src/main/java/com/linecorp/armeria/client/ClientFactoryBuilder.java [hunk 3]
```diff
@@ -300,6 +300,20 @@ public ClientFactoryBuilder tlsCustomizer(Consumer<? super SslContextBuilder> tl
         return this;
     }
 
+    /**
+     * Sets whether to allow
+     * <a href="https://tools.ietf.org/html/rfc7540#appendix-A">TLS 1.2 Cipher Suite Block List</a>
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/ClientFactoryBuilder.java:305]: Could you use disallowed instead of block?

Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/client/ClientFactoryBuilder.java:305]: `Sets whether to allow the bad cipher suites listed in <a href="...">RFC7540</a>.` ?

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/ClientFactoryBuilder.java:305]: Ah that's better 😄

Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/client/ClientFactoryBuilder.java:305]: Maybe just `Allows the bad cipher ...`

Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/client/ClientFactoryBuilder.java:305]: (and some warning message if we didn't add one yet)

File: core/src/main/java/com/linecorp/armeria/client/ClientFactoryBuilder.java [hunk 4]
```diff
@@ -300,6 +300,19 @@ public ClientFactoryBuilder tlsCustomizer(Consumer<? super SslContextBuilder> tl
         return this;
     }
 
+    /**
+     * Allow the bad cipher suites listed in
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/ClientFactoryBuilder.java:304]:
```suggestion
     * Allows the bad cipher suites listed in
```

File: core/src/main/java/com/linecorp/armeria/client/ClientFactoryBuilder.java [hunk 5]
```diff
@@ -300,6 +300,19 @@ public ClientFactoryBuilder tlsCustomizer(Consumer<? super SslContextBuilder> tl
...
+     * <a href="https://tools.ietf.org/html/rfc7540#appendix-A">RFC7540</a> for TLS handshake.
+     *
+     * <p>Note that the bad cipher suites do not meet minimum security requirements.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/ClientFactoryBuilder.java:307]: I'm not sure what 'meet minimum security requirements' means clearly. We should convey the message that explicitly discourages its use.

How about:

    <p>Note that enabling this option increases the security risk of your connection.
    Use it only when you must communicate with a legacy system that does not support
    secure cipher suites. See ... for more information

File: core/src/main/java/com/linecorp/armeria/client/ClientFactoryOptions.java [hunk 6]
```diff
@@ -78,6 +78,17 @@
     public static final ClientFactoryOption<Consumer<? super SslContextBuilder>> TLS_CUSTOMIZER =
             ClientFactoryOption.define("TLS_CUSTOMIZER", b -> { /* no-op */ });
 
+    /**
+     * Whether to allow the bad cipher suites listed in
+     * <a href="https://tools.ietf.org/html/rfc7540#appendix-A">RFC7540</a> for TLS handshake.
+     *
+     * <p>Note that the blocked cipher suites do not meet minimum security requirements.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/ClientFactoryOptions.java:85]: Ditto

File: core/src/main/java/com/linecorp/armeria/client/ClientFactoryOptions.java [hunk 7]
```diff
@@ -446,4 +457,16 @@ public MeterRegistry meterRegistry() {
     public ProxyConfigSelector proxyConfigSelector() {
         return get(PROXY_CONFIG_SELECTOR);
     }
+
+    /**
+     * Returns whether to allow the bad cipher suites listed in
+     * <a href="https://tools.ietf.org/html/rfc7540#appendix-A">RFC7540</a> for TLS handshake.
+     *
+     * <p>Note that the bad cipher suites do not meet minimum security requirements.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/ClientFactoryOptions.java:465]: Ditto

**Review by trustin — APPROVED** [review]
:bow:

**Pull request merged**